src/otx/algo/instance_segmentation/backbones/detectron_resnet.py

```diff
@@ -0,0 +1,332 @@
+        norm: str = "BN",
+        stride_in_1x1: bool = False,
+        dilation: int = 1,
+    ) -> None:
+        super().__init__(in_channels, out_channels, stride)
+
+        if in_channels != out_channels:
+            self.shortcut = Conv2d(
+                in_channels,
+                out_channels,
+                kernel_size=1,
+                stride=stride,
+                bias=False,
+                norm=get_norm(norm, out_channels),
+            )
+        else:
+            self.shortcut = None  # type: ignore[assignment]
+
+        # The original MSRA ResNet models have stride in the first 1x1 conv
+        # The subsequent fb.torch.resnet and Caffe2 ResNe[X]t implementations have
+        # stride in the 3x3 conv
+        stride_1x1, stride_3x3 = (stride, 1) if stride_in_1x1 else (1, stride)
+
+        self.conv1 = Conv2d(
+            in_channels,
+            bottleneck_channels,
+            kernel_size=1,
+            stride=stride_1x1,
+            bias=False,
+            norm=get_norm(norm, bottleneck_channels),
+        )
+
+        self.conv2 = Conv2d(
+            bottleneck_channels,
+            bottleneck_channels,
+            kernel_size=3,
+            stride=stride_3x3,
+            padding=1 * dilation,
+            bias=False,
+            groups=num_groups,
+            dilation=dilation,
+            norm=get_norm(norm, bottleneck_channels),
+        )
+
+        self.conv3 = Conv2d(
+            bottleneck_channels,
+            out_channels,
+            kernel_size=1,
+            bias=False,
+            norm=get_norm(norm, out_channels),
+        )
+
+        for layer in [self.conv1, self.conv2, self.conv3, self.shortcut]:
+            if layer is not None:  # shortcut can be None
+                c2_msra_fill(layer)
+
+    def forward(self, x: Tensor) -> Tensor:
+        """Forward pass."""
+        out = self.conv1(x)
+        out = f.relu_(out)
+
+        out = self.conv2(out)
+        out = f.relu_(out)
+
+        out = self.conv3(out)
+
+        shortcut = self.shortcut(x) if self.shortcut is not None else x
+
+        out += shortcut
+        return f.relu_(out)
+
+
+class BasicStem(CNNBlockBase):
+    """The standard ResNet stem (layers before the first residual block), with a conv, relu and max_pool."""
+
+    def __init__(self, in_channels: int = 3, out_channels: int = 64, norm: str = "BN") -> None:
+        super().__init__(in_channels, out_channels, 4)
+        self.in_channels = in_channels
+        self.conv1 = Conv2d(
+            in_channels,
+            out_channels,
+            kernel_size=7,
+            stride=2,
+            padding=3,
+            bias=False,
+            norm=get_norm(norm, out_channels),
+        )
+        c2_msra_fill(self.conv1)
+
+    def forward(self, x: Tensor) -> Tensor:
+        """Forward pass."""
+        x = self.conv1(x)
+        x = f.relu_(x)
+        return f.max_pool2d(x, kernel_size=3, stride=2, padding=1)
+
+
+class ResNet(nn.Module):
```

Comment:
Not at the moment. The structure of detectron.ResNet differs from the version of ResNet we have in MMDet, meaning we will have issues loading weights without implementing a function to handle the layer discrepancies. We can consider implementing this later, but it's not a priority right now.

Reply:
I think we must pursue to reuse the models implemented in otx.
If there are models redundantly used across tasks, it should be refactored as you said.
But for newly added models, adding ways to adapt them to otx models will reduce our future work.
Could you find ways for this?

Reply:
yeah, reuse is possible, but it's not exactly a striaghtforward task. It'd be great if everything is ready in one PR—but alas, I can only do it in stages. Rome wasn’t built in a day, yeah?

Reply:
Yes, it's annoying. But if we work on it while considering it, the new model will naturally fit into the library structure, the code will be cleaner, the PR size will be smaller, the person in charge will have less to consider, and the reviewers will have an easier time reviewing.

Reply:
I mean, using the model module directly from an existing module isn't straightforward. For instance, this PR is created before detection 3D, but after merging 3D PR first, we end up with duplicate code.

I think at this stage, the reviewer can be selective and focus on areas that impact OTX's core functionality, such as testing, exporting, prediction, and post-processing. Since much of the code is being migrated, it’s impossible to review every line in detail. Instead, we can leave the migrated code as is, as long as there are no major issues, and as implementer focusing on further refactoring.

This doesn’t mean we should migrate code without consideration, but rather that we can approach it in stages. I’m not sure a full model integration with complete refactoring is necessary all at once in one PR, and I’d be fine refactoring in a separate PR like we did for other PRs.